### PR TITLE
Implement an iofs.FS over the REAPI Tree proto

### PR DESCRIPTION
**Pull request by Tatskaari**

### CODE REVIEW

File: src/remote/fs/fs.go [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+	if fileNode != nil {
+		return fs.openFile(fileNode)
+	}
+	if dirNode != nil {
+		return fs.openDir(dirNode)
+	}
+	return nil, os.ErrNotExist
+}
+
+// openFile downloads a file from the CAS and returns it as an iofs.File
```
Comment (goddenrich) [inline src/remote/fs/fs.go:86]: I don't think it returns it as an iofs.File

Comment (goddenrich) [reply src/remote/fs/fs.go:90]: ah as in *file implements the iofs.File interface

File: src/remote/fs/fs.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+		info:     i.withProperties(dirPb.NodeProperties),
+		pb:       dirPb,
+		children: fs.directories,
+	}, nil
+}
+
+func (fs *CASFileSystem) findNode(wd *pb.Directory, name string) (*pb.FileNode, *pb.DirectoryNode, *pb.SymlinkNode, error) {
+	name, rest, hasToBeDir := strings.Cut(name, string(filepath.Separator))
+	// Must be a dodgy symlink that goes past our tree.
+	if name == ".." || name == "." {
```
Comment (goddenrich) [inline src/remote/fs/fs.go:124]: I don't think "." should be ErrNotExist

File: src/remote/fs/fs.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	for _, file := range p.pb.Files {
+		if n > 0 && len(ret) == n {
+			return ret, nil
+		}
+		ret = append(ret, &info{
+			name: file.Name,
+			mode: os.FileMode(file.NodeProperties.UnixMode.Value),
+			// Technically we could calculate this on demand by allowing info.Size() to download the file from the
+			// CAS... we don't need to for now though.
+			size: 0,
```
[the remark this answers is not in the document]
Comment (Tatskaari) [reply src/remote/fs/fs.go]: Ah yeah! It's on the digest! Nice spot.

File: src/remote/fs/fs.go [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+	return nil
+}
+
+type dir struct {
+	pb       *pb.Directory
+	children map[digest.Digest]*pb.Directory
+	*info
+}
+
+// ReadDir is a slightly incorrect implementation of ReadDir. It deviates slightly as it will report all files have 0
```
Comment (goddenrich) [inline src/remote/fs/fs.go:189]: comment should be updated now we have the filesize

**Review by goddenrich — APPROVED** [review]
Appart from

**Pull request merged**